apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -307,29 +341,35 @@ export async function aiGrade({
         null,
         [],
         question_course,
-        urls,
+        locals,
       );
-      if (render_question_results.courseIssues.length) {
+      if (render_question_results.courseIssues.length > 0) {
         job.info(render_question_results.courseIssues.toString());
         job.error('Error occurred');
         job.fail('Errors occurred while AI grading, see output for details');
       }
       const $ = cheerio.load(render_question_results.data.questionHtml, null, false);
       $('script').remove();
-      const question_prompt = $.html();
+      const questionPrompt = $.html();
 
-      const submission_embedding = await ensureSubmissionEmbedding({
-        submission_id: submission.id,
-        course,
-        question,
-        instance_question,
-        urlPrefix,
-        openai,
-      });
-      const student_answer = submission_embedding.submission_text;
+      let submission_embedding = await queryOptionalRow(
+        sql.select_embedding_for_submission,
+        { submission_id: submission.id },
+        SubmissionGradingContextEmbeddingSchema,
+      );
+      if (!submission_embedding) {
+        submission_embedding = await generateSubmissionEmbedding({
+          course,
+          question,
+          instance_question,
+          urlPrefix,
+          openai,
+        });
+      }
+      const submission_text = submission_embedding.submission_text;
 
       const example_submissions = await queryRows(
-        sql.select_closest_embeddings,
+        sql.select_closest_submission_info,
```

Comment:
At some point I think we should investigate how well this RAG is working. It's entirely possible that the default embedding we get isn't doing the best job and we should augment it with some fuzzy keywords or something else.